GPU support for multi precision build

oseikuffuor1 opened the pull request:
The goal of this PR is to enable building and testing mixed-precision Krylov solvers on GPUs.
* Fixed issues related to building with long double precision on GPUs
* Added utility functions such as array copy and clone to facilitate mixed-precision code development
* Added some new mixed-precision matrix and vector functions: clone, copy, convert functions
* Use hypre_double for dnum_nonzero member in hypre_parcsr_matrix struct and updated support for it within hypre.
* Added support for multiprecision build on GPU

oseikuffuor1 committed:
Merge branch 'master' into gpu-mixed-krylov

oseikuffuor1 committed:
Update autoconf build to prevent use of long double with GPUs.

oseikuffuor1 committed:
Add utility functions for mixed-precision function development.

oseikuffuor1 committed:
Add new mixed-precision matrix and vector functionality.

oseikuffuor1 committed:
Updates to support use of hypre_double for dnum_nonzeros in parcsr_matrix.

oseikuffuor1 committed:
Minor updates to test drivers.

rfalgout commented on src/config/HYPRE_config.h.in:
```diff
@@ -17,6 +17,9 @@
 #undef HAVE_INTTYPES_H
 
+/* Define to 1 if you have the <memory.h> header file. */
+#undef HAVE_MEMORY_H
```
Why is this needed?

rfalgout commented:
You may be aware of this already, but just in case...  This doesn't build cleanly on my mac currently.  There are a bunch of unused variable warnings, nothing major.

oseikuffuor1 committed:
Fixes for GPU build errors.

oseikuffuor1 commented:
> You may be aware of this already, but just in case... This doesn't build cleanly on my mac currently. There are a bunch of unused variable warnings, nothing major.

Thanks @rfalgout . Can you share which build options you are using?

rfalgout commented on src/utilities/HYPRE_utilities.h:
```diff
@@ -162,7 +162,11 @@ typedef HYPRE_Real HYPRE_Complex;
  * mixed precision code. */
 typedef double       hypre_double;
 typedef float        hypre_float;
+#if defined (HYPRE_USING_GPU)
```
Just curious.  Will this be true for all GPUs?

rfalgout commented on src/utilities/_hypre_utilities.hpp:
```diff
@@ -2475,6 +2490,8 @@ template <typename T>
                                                     HYPRE_Int *d_row_ptr, T *d_row_num, T *d_row_ind);
 
+template<typename T1, typename T2, typename T3>
+HYPRE_Int hypreDevice_Axpyzn_mp(HYPRE_Int n, T1 *d_x, T2 *d_y, T3 *d_z, T1 a, T2 b);
```
Why `hypreDevice_`?  There are other functions that append `Host` and `Device` just after the `Axpy` part (and similar for other functions).  Is this a GPU naming convention that I am just forgetting about?  I see there is also a `hypreFunctor_` function.

oseikuffuor1 replied:
My understanding and reason was what @victorapm said :-). I also agree it would be good to document this somewhere, so we are all on the same page.

oseikuffuor1 commented on src/parcsr_mv/HYPRE_parcsr_mv_mp.h:
```diff
@@ -7,8 +7,8 @@
- * Header file for HYPRE_parcsr_mv library
- *
+ * Header file for HYPRE_parcsr_mv librarys
+ *s
```
Yes I agree. That's a good point @rfalgout. I will try to add some of that now, so we don't get too lazy or forget when we start adding more functions. :-)

oseikuffuor1 commented on src/utilities/utilities.c:
```diff
@@ -265,3 +265,11 @@ hypre_ConvertIndicesToString(HYPRE_Int  size,
+/*--------------------------------------------------------------------------*
+* hypre_GetSizeOfPrecision: get size of precision type
+*--------------------------------------------------------------------------*/
+size_t hypre_GetSizeOfPrecision(void)
```
Sure, that makes sense.

oseikuffuor1 commented on src/seq_mv/_hypre_seq_mv.h:
```diff
@@ -465,6 +465,7 @@ HYPRE_Int hypre_CSRMatrixCheckSetNumNonzeros( hypre_CSRMatrix *matrix );
 HYPRE_Int hypre_CSRMatrixResize( hypre_CSRMatrix *matrix, HYPRE_Int new_num_rows,
                                  HYPRE_Int new_num_cols, HYPRE_Int new_num_nonzeros );
 HYPRE_Int hypre_CSRMatrixEliminateRowsCols(hypre_CSRMatrix *A, HYPRE_Int nrows, HYPRE_Int *rows);
+HYPRE_Int hypre_CSRMatrixResetData(hypre_CSRMatrix  *matrix);
```
Yes currently it is only used in the mixed-precision convert function. It provides a workaround for the ownership question, so perhaps we can revisit it later. It may also be useful if one needs to reset matrix data while keeping the structure unchanged?

oseikuffuor1 committed:
Cleanup typos and unnecessary print statements.

oseikuffuor1 committed:
Change utility function name GetSizeOfPrecision to GetSizeOfReal.

oseikuffuor1 committed:
Generate configure script with correct autoconf version.

oseikuffuor1 committed:
Cleanup unused variable warnings.

oseikuffuor1 commented on src/parcsr_mv/parcsr_mv_mp.c:
```diff
@@ -63,7 +63,8 @@ hypre_ParVectorConvert_mp( hypre_ParVector *v,
 /*--------------------------------------------------------------------------
- * Mixed-Precision hypre_ParVectorAxpy
+ * Mixed-precision matrix conversion
+ * Note: This converts only the diag and offd matrices
```
@ulrikeyang this comment was just to specify that we only convert the main csr matrices (diag and offdiag) in the hypre_ParCSRMatrix_struct struct. There are others like diagT and offdT, which I suppose we could convert if they are set (since we are changing the precision afterwards), but I assume they are currently unset and would be generated later.

ulrikeyang commented on src/seq_mv/seq_mv_mp.c:
```diff
@@ -22,199 +22,63 @@
-   return hypre_error_flag;
+   
+   /* Call mixed-precision axpy on vector data */
+   return hypre_RealArrayAxpyn_mp(hypre_VectorPrecision (x), xp, hypre_VectorPrecision (y), yp,
```
Is this really Axpyn ?

oseikuffuor1 committed:
Some fixes and cleanup and minor refactoring for GPU build.

oseikuffuor1 committed:
Add gpu function lists for utilities folder.

victorapm commented on src/utilities/device_utils.c:
```diff
@@ -3078,3 +3122,91 @@ hypre_bind_device( HYPRE_Int myid,
+   return (HYPRE_Int) MPI_Info_free(info);
+}
+#endif
+#endif
```
Just chiming in... I think these should probably go back to `mpistubs.c`. I see they were just being compiled for GPUs before, but I think it's fine to compile them always and just place together with all the other MPI wrapper calls

oseikuffuor1 replied:
Thanks @victorapm , I made the changes in commit [7163b69](https://github.com/hypre-space/hypre/pull/1423/commits/7163b696df273d3365c0a3b1980efaad2633fba4)

oseikuffuor1 committed:
Added GPU-specific function lists for remaining directories.

oseikuffuor1 committed:
Revert changes to MPI function location and remove unnecessary guards.

rfalgout committed:
Updating gen_code.sh script to build code for gpu functions

rfalgout committed:
Merge branch 'gpu-mixed-krylov' of github.com:hypre-space/hypre into gpu-mixed-krylov

oseikuffuor1 committed:
Fixed astyle changes in previous commits

rfalgout committed:
Merge branch 'gpu-mixed-krylov' of github.com:hypre-space/hypre into gpu-mixed-krylov

rfalgout committed:
Minor changes to config/gen_code.sh script

oseikuffuor1 committed:
Removed empty gpu function lists

oseikuffuor1 committed:
Minor edit to suppress warnings for directories with no GPU function lists.

rfalgout commented on src/config/generate_function_list.sh:
```diff
@@ -13,4 +13,6 @@
 # Use awk to avoid issues with spacing
-nm -P *.o *.obj | awk '$2 == "T" {print $1}' | sed -e 's/^_//' -e 's/_$//'
+# Demangle any c++ name mangling and filter _device_stub_ prefixes.
+nm -P *.o *.obj | awk '$2 == "T" {print $1}' | c++filt | sed -e 's/(.*$//' -e 's/^__device_stub__//' -e 's/^_//' -e 's/_$//'
```
Have you looked into the portability of c++flt?

oseikuffuor1 replied:
Yes, I have thought a bit about this. From what I can tell, it is supported in binutils, macOS, most BSD systems, MinGW/Cygwin and so should be fairly portable. Ideally we would use the demangle option the "-C" for nm, since it is built into the function, but unfortunately work well. There is also --demangle for nm on GNU/linux systems, but c++filt is more portable.

rfalgout committed:
Minor changes to gen_code.sh to make it easier to follow

rfalgout committed:
Cosmetic changes to gen_code.sh

rfalgout committed:
Changed mup.*.gpu to mup.*_gpu in utilities and modified the gen_code.sh script

rfalgout committed:
Renaming mup.*.gpu to mup.*_gpu

rfalgout committed:
Modified gen_code.sh script to write 'mup_*_gpu.c' files

rfalgout committed:
First draft of gen_code.sh to put the gpu functions in separate files

rfalgout committed:
Updating gen_code.sh to fix a couple of issues

rfalgout committed:
Added gpu if test to gen_code.sh script before writing '.hpp' file

oseikuffuor1 committed:
Minor edit to include internal header in wrapper function file.

oseikuffuor1 committed:
Changes to unify some vendor-specific return types.

oseikuffuor1 committed:
Generated new MuP wrapper code and headers for GPU support and some reformatting.

oseikuffuor1 committed:
Fixed linker issues and updated gen_code.sh script.

oseikuffuor1 committed:
Some minor re-org. of gen_code.sh script and gix for test driver.

oseikuffuor1 committed:
Merge branch 'master' into gpu-mixed-krylov

oseikuffuor1 committed:
Fixed build issues when default build precision is not double.

oseikuffuor1 committed:
Merge branch 'master' into gpu-mixed-krylov

oseikuffuor1 committed:
Updated to fix astyle warnings.

oseikuffuor1 committed:
Updates to fix check-license warnings.

oseikuffuor1 committed:
Merge branch 'master' into gpu-mixed-krylov

rfalgout commented on src/parcsr_mv/_hypre_parcsr_mv_mup_def.h:
```diff
@@ -333,5 +333,46 @@
 #define hypre_setparvectordataowner HYPRE_FIXEDPRECISION_FUNC ( hypre_setparvectordataowner )
 #define hypre_setparvectorrandomvalues HYPRE_FIXEDPRECISION_FUNC ( hypre_setparvectorrandomvalues )
 #define hypre_vectortoparvector HYPRE_FIXEDPRECISION_FUNC ( hypre_vectortoparvector )
+#define hypreGPUKernel_ConcatDiagAndOffd HYPRE_FIXEDPRECISION_FUNC ( hypreGPUKernel_ConcatDiagAndOffd )
```
Didn't we just change the names of these?

victorapm replied:
@liruipeng's PR has not been merged into master yet

rfalgout replied:
Yup.  I jumped the gun.  Thanks!

rfalgout commented on src/utilities/multiprecision.c:
```diff
@@ -11,7 +11,13 @@
+#elif defined(HYPRE_LONG_DOUBLE)
+HYPRE_Precision hypre__global_precision = HYPRE_REAL_LONGDOUBLE;
+#else //HYPRE_DOUBLE
 HYPRE_Precision hypre__global_precision = HYPRE_REAL_DOUBLE;
```
Can we just change `HYPRE_REAL_DOUBLE` here to `HYPRE_OBJECT_PRECISION`? See `HYPRE_utilities.h` and the "RDF" note related to the name.  It would be better to avoid replicating the if-else blocks.

rfalgout replied:
Ignore my previous comment - it's wrong.  We may want to consider defining something like `HYPRE_PRECISION` in `HYPRE_utilities.h` and using it here, again to avoid the if-else.  Just something to think about maybe.  Not important.

rfalgout replied:
Well, now I think my original comment may be right. :)  It's no biggee either way.  Let me think about it some more.  The more important thing is to get the regression tests all working.  Sorry for the rambling...

rfalgout committed:
Fixing a Makefile issue

rfalgout reviewed (approved):
This looks good to me!  I only tested the CPU code, but I assume Victor will find any issues there might be there.  Everything looked okay visually.  Thanks for all the hard work on this!

oseikuffuor1 commented:
Thanks @rfalgout!

victorapm committed:
Add GPU+Mixed precision support to CMake build (#1515)

This PR makes the mixed-precision build work consistently across
 - plain CPU CMake
 - CUDA CMake and autotools
 - HIP CMake
 - SYCL CMake and autotools
 - CUDA/HIP CMake with umpire
 - CUDA/HIP autotools with umpire

victorapm reviewed (approved):
Thanks for all the work, Daniel!

oseikuffuor1 committed:
Merge branch 'master' into gpu-mixed-krylov
